pkg/ruler/compat.go

```diff
@@ -143,10 +143,16 @@ func EngineQueryFunc(engine *promql.Engine, q storage.Queryable, overrides Rules
 	}
 }
 
-func MetricsQueryFunc(qf rules.QueryFunc, queries, failedQueries prometheus.Counter) rules.QueryFunc {
+func MetricsQueryFunc(qf rules.QueryFunc, queries, failedQueries prometheus.Counter, queryTime *prometheus.CounterVec, userID string) rules.QueryFunc {
```

Comment:
You can directly pass `prometheus.Counter()` instead of calling `queryTime.WithLabelValues(userID)` each time inside this function (calling `queryTime.WithLabelValues(userID)` just once is faster).

```suggestion
func MetricsQueryFunc(qf rules.QueryFunc, queries, failedQueries prometheus.Counter, queryTime prometheus.Counter) rules.QueryFunc {
```